ftl/node/builder.py

```diff
@@ -34,59 +35,84 @@
-        # So we build a hierarchy like:
-        # /$tmp/$destination_path/node_modules
-        # And use the -C flag to tar to root the tarball at /$tmp.
-        pkg_dir = tempfile.mkdtemp()
-        app_dir = os.path.join(pkg_dir, destination_path.strip("/"))
-        os.makedirs(app_dir)
-
-        # Copy out the relevant package descriptors to a tempdir.
-        for f in [_PACKAGE_LOCK, _PACKAGE_JSON]:
-            if self._ctx.Contains(f):
-                with open(os.path.join(app_dir, f), 'w') as w:
-                    w.write(self._ctx.GetFile(f))
-
-        check_gcp_build(json.loads(self._ctx.GetFile(_PACKAGE_JSON)), app_dir)
-        subprocess.check_call(
-            ['rm', '-rf', os.path.join(app_dir, 'node_modules')])
-        with ftl_util.Timing("npm_install"):
+    def GetBuildLayers(self):
+        descriptor_contents = ftl_util.descriptor_parser(
+            self.descriptor_files, self._ctx)
+        if descriptor_contents is None:
+            return [self.app_layer]
+        builder_lyrs = [
+            self.PackageLayer(self._ctx, None, self.descriptor_files,
+                              self.destination_path)
+        ]
+        builder_lyrs.append(self.app_layer)
+        return builder_lyrs
+
+    class PackageLayer(build_layer.BaseLayer):
+        def __init__(self, ctx, pkg_txt, descriptor_files, destination_path):
+            self._ctx = ctx
+            self._pkg_txt = pkg_txt
+            self._descriptor_files = descriptor_files
+            self._destination_path = destination_path
+
+        def GetCacheKey(self):
+            descriptor_contents = ftl_util.descriptor_parser(
+                self._descriptor_files, self._ctx)
+
+            return hashlib.sha256(descriptor_contents).hexdigest()
+
+        def BuildLayer(self):
+            """Override."""
+            overrides = self._generate_overrides()
+
+            layer, sha = self._gen_package_tar(self._pkg_txt,
+                                               self._destination_path)
+            logging.info('Generated layer with sha: %s', sha)
+            return layer, sha, overrides
+
+        def _generate_overrides(self):
+            pj_contents = {}
+            if self._ctx.Contains(_PACKAGE_JSON):
+                pj_contents = json.loads(self._ctx.GetFile(_PACKAGE_JSON))
+            entrypoint = parse_entrypoint(pj_contents)
+            return metadata.Overrides(
+                creation_time=str(datetime.date.today()) + "T00:00:00Z",
+                entrypoint=entrypoint)
+
+        def _gen_package_tar(self, pkg_txt, destination_path):
+            # Create temp directory to write package descriptor to
+            pkg_dir = tempfile.mkdtemp()
+            app_dir = os.path.join(pkg_dir, destination_path.strip("/"))
+            os.makedirs(app_dir)
+
+            # Copy out the relevant package descriptors to a tempdir.
+            ftl_util.descriptor_copy(self._ctx, self._descriptor_files,
```

Comment:
Should you move this util up onto the class itself? The class stores _ctx and _descriptor files already.

Reply:
Currently this is shared for both builder and layer class which makes it harder to add as a shared base class method as I want to split up the parsing logic or else I am passing cache key from builder to layer.  I think I will keep as is for now